# **TikTok Project**
**Course 2 - Get Started with Python**

Welcome to the TikTok Project!

You have just started as a data professional at TikTok.

The team is still in the early stages of the project. You have received notice that TikTok's leadership team has approved the project proposal. To gain clear insights to prepare for a claims classification model, TikTok's provided data must be examined to begin the process of exploratory data analysis (EDA).

A notebook was structured and prepared to help you in this project. Please complete the following questions.

# **Course 2 End-of-course project: Inspect and analyze data**

In this activity, you will examine data provided and prepare it for analysis.
<br/>

**The purpose** of this project is to investigate and understand the data provided. This activity will:

1.   Acquaint you with the data

2.   Compile summary information about the data

3.   Begin the process of EDA and reveal insights contained in the data

4.   Prepare you for more in-depth EDA, hypothesis testing, and statistical analysis

**The goal** is to construct a dataframe in Python, perform a cursory inspection of the provided dataset, and inform TikTok data team members of your findings.
<br/>
*This activity has three parts:*

**Part 1:** Understand the situation
* How can you best prepare to understand and organize the provided TikTok information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning and future exploratory data analysis (EDA) and statistical activities

* Compile summary information about the data to inform next steps

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into variables

<br/>

To complete the activity, follow the instructions and answer the questions below. Then, you will us your responses to these questions and the questions included in the Course 2 PACE Strategy Document to create an executive summary.

Be sure to complete this activity before moving on to Course 3. You can assess your work by comparing the results to a completed exemplar after completing the end-of-course project.

# **Identify data types and compile summary information**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

# **PACE stages**

<img src="images/Pace.png" width="100" height="100" align=left>

   *        [Plan](#scrollTo=psz51YkZVwtN&line=3&uniqifier=1)
   *        [Analyze](#scrollTo=mA7Mz_SnI8km&line=4&uniqifier=1)
   *        [Construct](#scrollTo=Lca9c8XON8lc&line=2&uniqifier=1)
   *        [Execute](#scrollTo=401PgchTPr4E&line=2&uniqifier=1)

<img src="images/Plan.png" width="100" height="100" align=left>


## **PACE: Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:



### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided information?


*Begin by exploring your dataset and consider reviewing the Data Dictionary.*

==> ENTER YOUR RESPONSE HERE

<img src="images/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Imports and data loading**

Start by importing the packages that you will need to load and explore the dataset. Make sure to use the following import statements:
*   `import pandas as pd`

*   `import numpy as np`


In [1]:
# Import packages
import numpy as np
import pandas as pd

Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [3]:
# Load dataset into dataframe
data = pd.read_csv("tiktok_dataset.csv")
print(type(data))

<class 'pandas.core.frame.DataFrame'>


### **Task 2b. Understand the data - Inspect the data**

View and inspect summary information about the dataframe by **coding the following:**

1. `data.head(10)`
2. `data.info()`
3. `data.describe()`

*Consider the following questions:*

**Question 1:** When reviewing the first few rows of the dataframe, what do you observe about the data? What does each row represent?

**Question 2:** When reviewing the `data.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Question 3:** When reviewing the `data.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values? Does it seem that there are outlier values?

















In [7]:
# Display and examine the first ten rows of the dataframe

data.head(10)

# claim_status    video_id  video_duration_sec  \
0   1        claim  7017666017                  59   
1   2        claim  4014381136                  32   
2   3        claim  9859838091                  31   
3   4        claim  1866847991                  25   
4   5        claim  7105231098                  19   
5   6        claim  8972200955                  35   
6   7        claim  4958886992                  16   
7   8        claim  2270982263                  41   
8   9        claim  5235769692                  50   
9  10        claim  4660861094                  45   

                            video_transcription_text verified_status  \
0  someone shared with me that drone deliveries a...    not verified   
1  someone shared with me that there are more mic...    not verified   
2  someone shared with me that american industria...    not verified   
3  someone shared with me that the metro of st. p...    not verified   
4  someone shared with me that the number of busi...    not verified   
5  someone shared with me that gross domestic pro...    not verified   
6  someone shared with me that elvis presley has ...    not verified   
7  someone shared with me that the best selling s...    not verified   
8  someone shared with me that about half of the ...    not verified   
9  someone shared with me that it would take a 50...        verified   

  author_ban_status  video_view_count  video_like_count  video_share_count  \
0      under review          343296.0           19425.0              241.0   
1            active          140877.0           77355.0            19034.0   
2            active          902185.0           97690.0             2858.0   
3            active          437506.0          239954.0            34812.0   
4            active           56167.0           34987.0             4110.0   
5      under review          336647.0          175546.0            62303.0   
6            active          750345.0          486192.0           193911.0   
7            active          547532.0            1072.0               50.0   
8            active           24819.0           10160.0             1050.0   
9            active          931587.0          171051.0            67739.0   

   video_download_count  video_comment_count  
0                   1.0                  0.0  
1                1161.0                684.0  
2                 833.0                329.0  
3                1234.0                584.0  
4                 547.0                152.0  
5                4293.0               1857.0  
6                8616.0               5446.0  
7                  22.0                 11.0  
8                  53.0                 27.0  
9                4104.0               2540.0

In [8]:
# Get summary info

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19382 entries, 0 to 19381
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   #                         19382 non-null  int64  
 1   claim_status              19084 non-null  object 
 2   video_id                  19382 non-null  int64  
 3   video_duration_sec        19382 non-null  int64  
 4   video_transcription_text  19084 non-null  object 
 5   verified_status           19382 non-null  object 
 6   author_ban_status         19382 non-null  object 
 7   video_view_count          19084 non-null  float64
 8   video_like_count          19084 non-null  float64
 9   video_share_count         19084 non-null  float64
 10  video_download_count      19084 non-null  float64
 11  video_comment_count       19084 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 1.8+ MB


In [10]:
# Get summary statistics

data.describe()

#      video_id  video_duration_sec  video_view_count  \
count  19382.000000  1.938200e+04        19382.000000      19084.000000   
mean    9691.500000  5.627454e+09           32.421732     254708.558688   
std     5595.245794  2.536440e+09           16.229967     322893.280814   
min        1.000000  1.234959e+09            5.000000         20.000000   
25%     4846.250000  3.430417e+09           18.000000       4942.500000   
50%     9691.500000  5.618664e+09           32.000000       9954.500000   
75%    14536.750000  7.843960e+09           47.000000     504327.000000   
max    19382.000000  9.999873e+09           60.000000     999817.000000   

       video_like_count  video_share_count  video_download_count  \
count      19084.000000       19084.000000          19084.000000   
mean       84304.636030       16735.248323           1049.429627   
std       133420.546814       32036.174350           2004.299894   
min            0.000000           0.000000              0.000000   
25%          810.750000         115.000000              7.000000   
50%         3403.500000         717.000000             46.000000   
75%       125020.000000       18222.000000           1156.250000   
max       657830.000000      256130.000000          14994.000000   

       video_comment_count  
count         19084.000000  
mean            349.312146  
std             799.638865  
min               0.000000  
25%               1.000000  
50%               9.000000  
75%             292.000000  
max            9599.000000

===> ENTER YOUR RESPONSE TO QUESTIONS 1-3 HERE

**Question 1:** When reviewing the first few rows of the dataframe, what do you observe about the data? What does each row represent?

#### A: The rows have different data types like str, int and floats. One can see the duration of each video, the number of likes and the number of times the video has been shared, commented and downloaded.

**Question 2:** When reviewing the `data.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

#### A: One can see that some rows have all their entries (19382 non-null)while some others have null entries (19084 non-null, for view count for instance). There is difference of ~300 observations that do not have enties (null entries), such variables are clam_status, video_transcription_text, and the video_* variables.

**Question 3:** When reviewing the `data.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values? Does it seem that there are outlier values?

#### A: One can see that the standard deviation is pretty large, this shows that the data is signigicantly spead out. This is mostly because some videos have 0 likes, 0 commnets, 0 downloads, while others have hundreds of thousands.

### **Task 2c. Understand the data - Investigate the variables**

In this phase, you will begin to investigate the variables more closely to better understand them.

You know from the project proposal that the ultimate objective is to use machine learning to classify videos as either claims or opinions. A good first step towards understanding the data might therefore be examining the `claim_status` variable. Begin by determining how many videos there are for each different claim status.

In [31]:
# What are the different values for claim status and how many of each are in the data?

print(data['claim_status'].count())
data['claim_status'].value_counts()

19084


claim      9608
opinion    9476
Name: claim_status, dtype: int64

**Question:** What do you notice about the values shown?

#### A: total number of counts is 19084 i.e. 9608(claim) + 9476(opinion) = 19084

Next, examine the engagement trends associated with each different claim status.

Start by using Boolean masking to filter the data according to claim status, then calculate the mean and median view counts for each claim status.

In [46]:
# What is the average view count of videos with "claim" status?

mask = data['claim_status'] == 'claim'
data_claim = data[mask]

mask = data['claim_status'] == 'opinion'
data_opinion = data[mask]
data_opinion.head()

# claim_status    video_id  video_duration_sec  \
9608  9609      opinion  2200945278                  32   
9609  9610      opinion  1670297626                  22   
9610  9611      opinion  4356008348                  18   
9611  9612      opinion  4285230093                  39   
9612  9613      opinion  5927475305                  45   

                               video_transcription_text verified_status  \
9608      i think that humans will live on mars by 2035    not verified   
9609  i think that we will all fly with individual j...        verified   
9610  i think that drone deliveries are already happ...        verified   
9611  i think that there are more microorganisms in ...    not verified   
9612  i think that dubai is an island created by humans    not verified   

     author_ban_status  video_view_count  video_like_count  video_share_count  \
9608            active            5433.0            1602.0              381.0   
9609            active            6748.0            2373.0              387.0   
9610            active            5890.0            1501.0              580.0   
9611            active             642.0              49.0                5.0   
9612            active            7185.0            1771.0              577.0   

      video_download_count  video_comment_count  
9608                  12.0                  1.0  
9609                  37.0                  4.0  
9610                  24.0                  1.0  
9611                   0.0                  0.0  
9612                  28.0                  0.0

In [117]:
# What is the average view count of videos with "opinion" status?

print("mean view count opinions : ", data_opinion['video_view_count'].mean())
print("median view count opinions : ", data_opinion['video_view_count'].median())
print()
print("mean view count claims : ", data_claim['video_view_count'].mean())
print("median view count claims : ", data_claim['video_view_count'].median())

mean view count opinions :  4956.43224989447
median view count opinions :  4953.0

mean view count claims :  501029.4527477102
median view count claims :  501555.0


**Question:** What do you notice about the mean and media within each claim category?

#### A: for the opinions the median and mean are close to eachother.

Now, examine trends associated with the ban status of the author.

Use `groupby()` to calculate how many videos there are for each combination of categories of claim status and author ban status.

In [68]:
# Get counts for each group combination of claim status and author ban status
### YOUR CODE HERE ###

data.groupby(['claim_status','author_ban_status']).count()

#  video_id  video_duration_sec  \
claim_status author_ban_status                                       
claim        active             6566      6566                6566   
             banned             1439      1439                1439   
             under review       1603      1603                1603   
opinion      active             8817      8817                8817   
             banned              196       196                 196   
             under review        463       463                 463   

                                video_transcription_text  verified_status  \
claim_status author_ban_status                                              
claim        active                                 6566             6566   
             banned                                 1439             1439   
             under review                           1603             1603   
opinion      active                                 8817             8817   
             banned                                  196              196   
             under review                            463              463   

                                video_view_count  video_like_count  \
claim_status author_ban_status                                       
claim        active                         6566              6566   
             banned                         1439              1439   
             under review                   1603              1603   
opinion      active                         8817              8817   
             banned                          196               196   
             under review                    463               463   

                                video_share_count  video_download_count  \
claim_status author_ban_status                                            
claim        active                          6566                  6566   
             banned                          1439                  1439   
             under review                    1603                  1603   
opinion      active                          8817                  8817   
             banned                           196                   196   
             under review                     463                   463   

                                video_comment_count  
claim_status author_ban_status                       
claim        active                            6566  
             banned                            1439  
             under review                      1603  
opinion      active                            8817  
             banned                             196  
             under review                       463

**Question:** What do you notice about the number of claims videos with banned authors? Why might this relationship occur?

#### A: the number of claims with banned authors is 1439 while the number of opinions with banned authors is reletively lower 196. This is because claims need to be fact-checked and thus tend to get the authors banned more easily idf they violate TikTok's rules.

Continue investigating engagement levels, now focusing on `author_ban_status`.

Calculate the median video share count of each author ban status.

In [87]:
# What's the median video share count of each author ban status?

mask = data['author_ban_status'] == 'active'
data_active = data[mask]
print(data_active['video_share_count'].median())

mask = data['author_ban_status'] == 'banned'
data_banned = data[mask]
print(data_banned['video_share_count'].median())

mask = data['author_ban_status'] == 'under review'
data_under_review = data[mask]
print(data_under_review['video_share_count'].median())

437.0
14468.0
9444.0


**Question:** What do you notice about the share count of banned authors, compared to that of active authors? 
Explore this in more depth.

#### A: the video share count of banned authors is the highest in average 14489, followed by the ones under review 9444, followed by the ones active 437. People share more opinions/claims by banned authors.

Use `groupby()` to group the data by `author_ban_status`, then use `agg()` to get the count, mean, and median of each of the following columns:
* `video_view_count`
* `video_like_count`
* `video_share_count`

Remember, the argument for the `agg()` function is a dictionary whose keys are columns. The values for each column are a list of the calculations you want to perform.

In [121]:
#data.groupby('author_ban_status').agg(['count','mean','median'])

data.groupby(['claim_status','author_ban_status']).agg(
    ['count','mean','median'])[['video_view_count','video_like_count','video_share_count']] 

video_view_count                           \
                                          count           mean    median   
claim_status author_ban_status                                             
claim        active                        6566  499221.733171  499500.5   
             banned                        1439  505907.917304  512572.0   
             under review                  1603  504054.640674  500774.0   
opinion      active                        8817    4958.120563    4958.0   
             banned                         196    4876.530612    5083.5   
             under review                   463    4958.105832    4884.0   

                               video_like_count                           \
                                          count           mean    median   
claim_status author_ban_status                                             
claim        active                        6566  164960.302924  121943.0   
             banned                        1439  173719.102849  132044.0   
             under review                  1603  165566.954460  125882.0   
opinion      active                        8817    1091.714982     820.0   
             banned                         196    1027.515306     799.5   
             under review                   463    1139.663067     876.0   

                               video_share_count                         
                                           count          mean   median  
claim_status author_ban_status                                           
claim        active                         6566  32769.101889  17774.5  
             banned                         1439  34056.580959  19018.0  
             under review                   1603  33155.623206  18084.0  
opinion      active                         8817    217.166950    121.0  
             banned                          196    208.423469    108.5  
             under review                    463    220.431965    124.0

**Question:** What do you notice about the number of views, likes, and shares for banned authors compared to active authors?

#### A: For banned we have in average 445845 views, 153017 likes, 29998 shares. One the other hand, for active usuers we have in average 215927 views, 71036 likes, 14111 shares. Banned users have almost 2.0 times more views, 2.15 times more likes and 2.12 more shares than active users.

Now, create three new columns to help better understand engagement rates:
* `likes_per_view`: represents the number of likes divided by the number of views for each video
* `comments_per_view`: represents the number of comments divided by the number of views for each video
* `shares_per_view`: represents the number of shares divided by the number of views for each video

In [105]:
# Create a likes_per_view column
data['likes_per_view'] = data['video_like_count']/data['video_view_count']

# Create a comments_per_view column
data['comments_per_view'] = data['video_comment_count']/data['video_view_count']

# Create a shares_per_view column
data['shares_per_view'] = data['video_share_count']/data['video_view_count']

data.sort_values(by='likes_per_view', ascending=False)

# claim_status    video_id  video_duration_sec  \
3638    3639        claim  7327101059                  41   
8892    8893        claim  8261213140                  34   
5109    5110        claim  6821021715                  16   
433      434        claim  5865719227                  34   
5711    5712        claim  7572100502                  19   
...      ...          ...         ...                 ...   
19377  19378          NaN  7578226840                  21   
19378  19379          NaN  6079236179                  53   
19379  19380          NaN  2565539685                  10   
19380  19381          NaN  2969178540                  24   
19381  19382          NaN  8132759688                  13   

                                video_transcription_text verified_status  \
3638   someone learned  from the media that violins a...    not verified   
8892   someone discovered on the news a claim that fr...    not verified   
5109   i learned  from the media that the folds in a ...    not verified   
433    a friend claimed someone revealed that the lon...    not verified   
5711   i read  in the news that the band metallica is...    not verified   
...                                                  ...             ...   
19377                                                NaN    not verified   
19378                                                NaN    not verified   
19379                                                NaN        verified   
19380                                                NaN    not verified   
19381                                                NaN    not verified   

      author_ban_status  video_view_count  video_like_count  \
3638             active          569443.0          379618.0   
8892             active          306060.0          204007.0   
5109             active          241233.0          160772.0   
433              active          389440.0          259544.0   
5711             active          882170.0          587724.0   
...                 ...               ...               ...   
19377            active               NaN               NaN   
19378            active               NaN               NaN   
19379      under review               NaN               NaN   
19380            active               NaN               NaN   
19381            active               NaN               NaN   

       video_share_count  video_download_count  video_comment_count  \
3638             33175.0                1207.0                801.0   
8892             78365.0                1518.0                783.0   
5109             32631.0                2652.0               1072.0   
433              77289.0                5242.0               1128.0   
5711            234618.0                2409.0                494.0   
...                  ...                   ...                  ...   
19377                NaN                   NaN                  NaN   
19378                NaN                   NaN                  NaN   
19379                NaN                   NaN                  NaN   
19380                NaN                   NaN                  NaN   
19381                NaN                   NaN                  NaN   

       likes_per_view  comments_per_view  shares_per_view  
3638         0.666648           0.001407         0.058259  
8892         0.666559           0.002558         0.256045  
5109         0.666459           0.004444         0.135268  
433          0.666454           0.002896         0.198462  
5711         0.666225           0.000560         0.265956  
...               ...                ...              ...  
19377             NaN                NaN              NaN  
19378             NaN                NaN              NaN  
19379             NaN                NaN              NaN  
19380             NaN                NaN              NaN  
19381             NaN                NaN              NaN  

[19382 rows x 15 columns]

Use `groupby()` to compile the information in each of the three newly created columns for each combination of categories of claim status and author ban status, then use `agg()` to calculate the count, the mean, and the median of each group.

In [122]:
grouped = data.groupby(['claim_status']).agg(['count','mean','median'])
grouped[['likes_per_view','comments_per_view','shares_per_view']]

likes_per_view                     comments_per_view            \
                      count      mean    median             count      mean   
claim_status                                                                  
claim                  9608  0.331610  0.329734              9608  0.001386   
opinion                9476  0.219802  0.218127              9476  0.000516   

                       shares_per_view                      
                median           count      mean    median  
claim_status                                                
claim         0.000774            9608  0.065867  0.049673  
opinion       0.000253            9476  0.043699  0.032487

**Question:**

How does the data for claim videos and opinion videos compare or differ? Consider views, comments, likes, and shares.
#### A: roughly speaking the likes, comments and shares per view from claims are in average almost twice than those from opinions. Particulary, the shares per view of claims is 1.3 times larger than those of opinions. The comments per views pf claims are 2.6 times larger than those of opinions. And the likes per view of claims are 1.5 times larger than those of opinions.

<img src="images/Construct.png" width="100" height="100" align=left>

## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.




<img src="images/Execute.png" width="100" height="100" align=left>

## **PACE: Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response.

### **Given your efforts, what can you summarize for Rosie Mae Bradshaw and the TikTok data team?**

*Note for Learners: Your answer should address TikTok's request for a summary that covers the following points:*

*   What percentage of the data is comprised of claims and what percentage is comprised of opinions?
*   What factors correlate with a video's claim status?
*   What factors correlate with a video's engagement level?


#### A: We have that 9608 are claims while 9476 are opinins, thus almost 50% are almost claims, 50% opinions. The factors that correlate with a video's claim status is the authors banned status: videos from banned users have in average more likes, comments and shares. Because these videos are claims, they tend to get banned more often.

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.